EIPS/eip-custom_errors_for_erc_tokens.md

```diff
@@ -0,0 +1,504 @@
+
+At the moment of the release of custom errors, the standard tokens ([EIP-20](./eip-20.md), [EIP-721](./eip-721.md), [EIP-1155](./eip-1155.md)) were already in finalized state, so no error specification was included.
+
+An error specification will allow users to expect more consistent error messages across applications or testing environments, while exposing pertinent arguments and overall reducing the need of writing expensive revert strings in the deployment bytecode.
+
+The list of token standard errors is defined in the specification below.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+The provided interfaces with errors SHOULD conform with the [error grammar rules](#error-grammar-rules) described in the [rationale](#rationale) section.
+
+It's important to note that errors MAY be thrown in each scenario, but that's up to the implementers, unless a revert reason is specified in their corresponding EIPs.
+
+This EIP aims to provide a clear standard only in case an error will be thrown.
+
+### [EIP-20](./eip-20.md)
+
+#### `ERC20InvalidSender(address)`
+
+Indicates a failure with the token sender.
+
+##### Parameters
+
+- `address _sender`: The address of whose tokens are transferred from.
+
+Throw when the cause of failure is the sender in a transfer.
+
+- MUST NOT be thrown for approval operations.
+- MUST NOT be thrown for balance or allowance requirements.
+  - Use `ERC20InsufficientBalance` or `ERC20InsufficientAllowance` instead.
+- MUST be thrown for unintended transfers from the zero address.
+
+#### `ERC20InvalidReceiver(address)`
+
+Indicates a failure with the token receiver.
+
+##### Parameters
+
+- `address _receiver` The address of whose tokens are transferred to.
+
+Throw when the cause of failure is the receiver in a transfer.
+
+- MUST NOT be thrown for approval operations.
+- MUST be thrown for unintended transfers to the zero address.
+- MUST be thrown for unintended transfers to non-compatible addresses (eg. contract addresses).
+
+#### `ERC20InsufficientBalance(address, uint256, uint256)`
+
+Indicates an error related to the current `_balance` of a sender in a transfer.
+
+##### Parameters
+
+- `address _sender`: The address of whose tokens are transferred from.
+- `uint256 _balance`: The current amount of token the sender has.
+- `uint256 _needed`: The amount of token the sender needs to perform the operation.
+
+Throw when the cause of failure is the balance of a sender.
+
+- MUST NOT be thrown if `_balance` is equal or greater than `_needed`.
+- MUST be thrown when `_balance` is less than `_needed`.
+
+#### `ERC20InvalidApprover(address)`
```

Comment:
I think we need to leave this one out of the spec, even though we might use it in OpenZeppelin Contracts. It would never show up in a standard ERC20 token because the approver/owner is always `msg.sender`.

Reply:
I was about to say the same in your edit.
We don't know the specific use case of an ERC20. Maybe an approver might be blacklisted and we can't know it or enforce otherwise. I agree we're providing a clear go-to when those cases happen.